File: AndroidApp/app/src/test/java/com/acme/taskflow/SyncSquashingTest.kt
package com.acme.taskflow

import com.acme.taskflow.data.LocalSyncOperation
import com.acme.taskflow.data.SyncSquashing
import com.acme.taskflow.data.child
import com.acme.taskflow.data.json
import com.acme.taskflow.data.number
import com.acme.taskflow.data.obj
import com.acme.taskflow.data.text
import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.UUID

class SyncSquashingTest {

    @Test
    fun putPlusPut_mergesPayloads_and_preservesEarliestExpectedVersion() {
        val list = mutableListOf<LocalSyncOperation>()
        val orgId = UUID.randomUUID().toString()
        val taskId = UUID.randomUUID().toString()

        val olderPayload = json("title" to "Old Title", "expectedVersion" to 10)
        val newerPayload = json("priority" to "high", "expectedVersion" to 999)

        val op1 = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "PUT",
            payloadJson = olderPayload.toString(),
            dirtyFields = listOf("title")
        )
        val op2 = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "PUT",
            payloadJson = newerPayload.toString(),
            dirtyFields = listOf("priority")
        )

        SyncSquashing.squash(list, op1)
        val squashed = SyncSquashing.squash(list, op2)

        assertTrue(squashed)
        assertEquals(1, list.size)
        assertEquals("PUT", list[0].operation)

        val merged = JsonParser.parseString(list[0].payloadJson!!).obj()
        assertEquals("Old Title", merged.text("title"))
        assertEquals("high", merged.text("priority"))
        // Preserves earliest expectedVersion
        assertEquals(10, merged.number("expectedVersion"))
        assertEquals(listOf("priority", "title"), list[0].dirtyFields)
    }

    @Test
    fun postPlusPut_foldsIntoCreate() {
        val list = mutableListOf<LocalSyncOperation>()
        val orgId = UUID.randomUUID().toString()
        val taskId = UUID.randomUUID().toString()

        val createPayload = json("id" to taskId, "title" to "Initial Title", "list_id" to "list-1")
        val updatePayload = json("title" to "Updated Title", "description" to "Some details", "expectedVersion" to 1)

        val opPost = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "POST",
            payloadJson = createPayload.toString()
        )
        val opPut = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "PUT",
            payloadJson = updatePayload.toString(),
            dirtyFields = listOf("title", "description")
        )

        SyncSquashing.squash(list, opPost)
        val squashed = SyncSquashing.squash(list, opPut)

        assertTrue(squashed)
        assertEquals(1, list.size)
        assertEquals("POST", list[0].operation)

        val folded = JsonParser.parseString(list[0].payloadJson!!).obj()
        assertEquals(taskId, folded.text("id"))
        assertEquals("Updated Title", folded.text("title"))
        assertEquals("Some details", folded.text("description"))
        assertEquals("list-1", folded.text("list_id"))
        assertFalse(folded.has("expectedVersion"))
    }

    @Test
    fun putPlusDelete_replacesWithDelete() {
        val list = mutableListOf<LocalSyncOperation>()
        val orgId = UUID.randomUUID().toString()
        val taskId = UUID.randomUUID().toString()

        val opPut = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "PUT",
            payloadJson = json("title" to "Updated").toString()
        )
        val opDelete = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "DELETE"
        )

        SyncSquashing.squash(list, opPut)
        val squashed = SyncSquashing.squash(list, opDelete)

        assertTrue(squashed)
        assertEquals(1, list.size)
        assertEquals("DELETE", list[0].operation)
    }

    @Test
    fun postPlusDelete_becomesNoop() {
        val list = mutableListOf<LocalSyncOperation>()
        val orgId = UUID.randomUUID().toString()
        val taskId = UUID.randomUUID().toString()

        val opPost = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "POST",
            payloadJson = json("title" to "Offline Created").toString()
        )
        val opDelete = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "DELETE"
        )

        SyncSquashing.squash(list, opPost)
        val squashed = SyncSquashing.squash(list, opDelete)

        assertTrue(squashed)
        assertTrue(list.isEmpty())
    }

    @Test
    fun deletePlusPut_marksNeedsAttention() {
        val list = mutableListOf<LocalSyncOperation>()
        val orgId = UUID.randomUUID().toString()
        val taskId = UUID.randomUUID().toString()

        val opDelete = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "DELETE"
        )
        val opPut = LocalSyncOperation(
            entityType = "task",
            entityId = taskId,
            orgId = orgId,
            operation = "PUT",
            payloadJson = json("title" to "Edit after delete").toString()
        )

        SyncSquashing.squash(list, opDelete)
        val squashed = SyncSquashing.squash(list, opPut)

        assertTrue(squashed)
        assertEquals(1, list.size)
        assertTrue(list[0].needsAttention)
        assertTrue(list[0].lastError?.contains("Invalid offline sequence") == true)
    }
}

File: WebApp/test/syncSquashing.test.js
import { test } from 'node:test';
import assert from 'node:assert';

// Standalone function matching squashSyncOperations in syncEngine.ts
function squashSyncOperations(queue, incoming) {
  let existingIdx = -1;
  for (let i = queue.length - 1; i >= 0; i--) {
    const op = queue[i];
    if (
      op.orgId === incoming.orgId &&
      op.entityType === incoming.entityType &&
      op.entityId === incoming.entityId &&
      !op.needsAttention
    ) {
      existingIdx = i;
      break;
    }
  }

  if (existingIdx < 0) {
    queue.push(incoming);
    return false;
  }

  const last = queue[existingIdx];
  const lastOp = last.operation;
  const incOp = incoming.operation;

  // Rule 1: PUT + PUT -> merge fields, preserve earliest expectedVersion
  if (lastOp === 'PUT' && incOp === 'PUT') {
    const older = { ...(last.payload || {}) };
    const newer = { ...(incoming.payload || {}) };

    const earliestVersion =
      older.expectedVersion !== undefined
        ? older.expectedVersion
        : newer.expectedVersion;

    const mergedPayload = { ...older, ...newer };
    if (earliestVersion !== undefined) {
      mergedPayload.expectedVersion = earliestVersion;
    }

    last.payload = mergedPayload;
    last.dirtyFields = Array.from(
      new Set([...(last.dirtyFields || []), ...(incoming.dirtyFields || [])])
    ).sort();
    last.timestamp = Math.max(last.timestamp, incoming.timestamp);
    return true;
  }

  // Rule 2: POST + PUT -> fold updates directly into create POST
  if (lastOp === 'POST' && incOp === 'PUT') {
    const create = { ...(last.payload || {}) };
    const update = { ...(incoming.payload || {}) };
    delete update.expectedVersion;

    last.payload = { ...create, ...update };
    last.timestamp = Math.max(last.timestamp, incoming.timestamp);
    return true;
  }

  // Rule 3: PUT + DELETE -> replace PUT with single DELETE
  if (lastOp === 'PUT' && incOp === 'DELETE') {
    queue.splice(existingIdx, 1, incoming);
    return true;
  }

  // Rule 4: POST + DELETE -> created and deleted offline => cancel both (NO-OP)
  if (lastOp === 'POST' && incOp === 'DELETE') {
    queue.splice(existingIdx, 1);
    return true;
  }

  // Rule 5: DELETE + PUT -> invalid offline sequence
  if (lastOp === 'DELETE' && incOp === 'PUT') {
    last.needsAttention = true;
    last.lastError = 'Invalid offline sequence: DELETE followed by PUT.';
    return true;
  }

  queue.push(incoming);
  return false;
}

test('putPlusPut_mergesPayloads_and_preservesEarliestExpectedVersion', () => {
  const queue = [];
  const orgId = 'org-1';
  const taskId = 'task-1';

  const op1 = {
    id: 'op-1',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'PUT',
    payload: { title: 'Old Title', expectedVersion: 10 },
    dirtyFields: ['title'],
    timestamp: 1000,
  };
  const op2 = {
    id: 'op-2',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'PUT',
    payload: { priority: 'high', expectedVersion: 999 },
    dirtyFields: ['priority'],
    timestamp: 2000,
  };

  squashSyncOperations(queue, op1);
  const squashed = squashSyncOperations(queue, op2);

  assert.strictEqual(squashed, true);
  assert.strictEqual(queue.length, 1);
  assert.strictEqual(queue[0].operation, 'PUT');
  assert.strictEqual(queue[0].payload.title, 'Old Title');
  assert.strictEqual(queue[0].payload.priority, 'high');
  assert.strictEqual(queue[0].payload.expectedVersion, 10);
  assert.deepStrictEqual(queue[0].dirtyFields, ['priority', 'title']);
});

test('postPlusPut_foldsIntoCreate', () => {
  const queue = [];
  const orgId = 'org-1';
  const taskId = 'task-1';

  const opPost = {
    id: 'op-post',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'POST',
    payload: { id: taskId, title: 'Initial Title', listId: 'list-1' },
    timestamp: 1000,
  };
  const opPut = {
    id: 'op-put',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'PUT',
    payload: { title: 'Updated Title', description: 'Some details', expectedVersion: 1 },
    dirtyFields: ['title', 'description'],
    timestamp: 2000,
  };

  squashSyncOperations(queue, opPost);
  const squashed = squashSyncOperations(queue, opPut);

  assert.strictEqual(squashed, true);
  assert.strictEqual(queue.length, 1);
  assert.strictEqual(queue[0].operation, 'POST');
  assert.strictEqual(queue[0].payload.title, 'Updated Title');
  assert.strictEqual(queue[0].payload.description, 'Some details');
  assert.strictEqual(queue[0].payload.listId, 'list-1');
  assert.strictEqual(queue[0].payload.expectedVersion, undefined);
});

test('putPlusDelete_replacesWithDelete', () => {
  const queue = [];
  const orgId = 'org-1';
  const taskId = 'task-1';

  const opPut = {
    id: 'op-put',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'PUT',
    payload: { title: 'Updated' },
    timestamp: 1000,
  };
  const opDelete = {
    id: 'op-delete',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'DELETE',
    timestamp: 2000,
  };

  squashSyncOperations(queue, opPut);
  const squashed = squashSyncOperations(queue, opDelete);

  assert.strictEqual(squashed, true);
  assert.strictEqual(queue.length, 1);
  assert.strictEqual(queue[0].operation, 'DELETE');
});

test('postPlusDelete_becomesNoop', () => {
  const queue = [];
  const orgId = 'org-1';
  const taskId = 'task-1';

  const opPost = {
    id: 'op-post',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'POST',
    payload: { title: 'Offline Created' },
    timestamp: 1000,
  };
  const opDelete = {
    id: 'op-delete',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'DELETE',
    timestamp: 2000,
  };

  squashSyncOperations(queue, opPost);
  const squashed = squashSyncOperations(queue, opDelete);

  assert.strictEqual(squashed, true);
  assert.strictEqual(queue.length, 0);
});

test('deletePlusPut_marksNeedsAttention', () => {
  const queue = [];
  const orgId = 'org-1';
  const taskId = 'task-1';

  const opDelete = {
    id: 'op-delete',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'DELETE',
    timestamp: 1000,
  };
  const opPut = {
    id: 'op-put',
    entityType: 'task',
    entityId: taskId,
    orgId,
    operation: 'PUT',
    payload: { title: 'Zombie Edit' },
    timestamp: 2000,
  };

  squashSyncOperations(queue, opDelete);
  const squashed = squashSyncOperations(queue, opPut);

  assert.strictEqual(squashed, true);
  assert.strictEqual(queue.length, 1);
  assert.strictEqual(queue[0].needsAttention, true);
  assert.ok(queue[0].lastError.includes('Invalid offline sequence'));
});
